crates/db_views/src/post_view.rs

```diff
@@ -417,12 +429,43 @@ impl PostView {
   }
 }
 
-#[derive(Default)]
+impl PaginationCursor {
+  // get cursor for page after the given posts
+  pub fn after(posts: &[PostView]) -> Option<PaginationCursor> {
+    posts
+      .last()
+      .map(|p| PaginationCursor(format!("P{:x}", p.post.id.0)))
+  }
+  pub async fn read(&self, pool: &mut DbPool<'_>) -> Result<PaginationCursorData, Error> {
+    Ok(PaginationCursorData(
+      PostAggregates::read(
+        pool,
+        PostId(
+          self
+            .0
+            .get(1..)
+            .and_then(|e| i32::from_str_radix(e, 16).ok())
+            .ok_or_else(|| Error::QueryBuilderError("Could not parse pagination token".into()))?,
+        ),
+      )
+      .await?,
+    ))
+  }
+}
+
+// currently we use a postaggregates struct as the pagination token.
+// we only use some of the properties of the post aggregates, depending on which sort type we page by
+#[derive(Clone)]
+pub struct PaginationCursorData(PostAggregates);
+
+#[derive(Default, Clone)]
 pub struct PostQuery<'a> {
   pub listing_type: Option<ListingType>,
   pub sort: Option<SortType>,
   pub creator_id: Option<PersonId>,
   pub community_id: Option<CommunityId>,
+  // if true, the query should be handled as if community_id was not given except adding the literal filter
+  pub community_id_just_for_prefetch: bool,
```

Comment:
Why can't you just check the `community_id.is_some()` ?

Reply:
because there's three separate cases:

1. user requests posts for a single community. then we need to use `featured_community` for sort+filter
2. user requests posts for multiple communities. the query is the main fetch. then we need to use `featured_local` for sort+filter
3. user requests post for multiple communities. the query is the prefetch. then we also need to use `featured_local` even though community_id is set to a value.

as to why a prefetch is needed see the second chart i posted in the comments

Reply:
Reminder... featured_community already has a dedicated TRIGGER that fires when it is changed. We could eliminate it from the sort order by playing a trick of setting the posted date + 10 years on a featured post. This is a worthwhile optimization to consider. Or we could manipulate the score and/or hot_rank values.

Reply:
There are a lot of filter possibilities, it could get extremely complex to do that approach for all of them.

Reply:
This doesn't have to be done for every filter or filter combination, just for any filter where there is an unbounded subset of sets of posts we need to fetch. Right now, the only case where it's needed is user followed communities. Another thing where it would be required is filtering by e.g. user followed instances or user followed users.

In addition, it only needs to be done when the filters are *implicitly* cross-correlated (in this case, the determination of whether we want featured_community or featured_local depends on whether we also filter by community_id)

I can't really think of any other instance where we would need to do this actually except for the already implemented community_id case. It's a bit of a hack to be sure, but it's the easiest way to prevent code duplication.

The reason is that it's only needed for the case where postgresql has to fetch multiple different ordered segments from one index. E.g. it does not apply to the all fetch or to explicit filters like local=true or banned=false.